cupy/core/_gufuncs.py

```diff
@@ -0,0 +1,500 @@
+        )
+
+        if len(input_coredimss) != len(args):
+            ValueError(
+                'According to `signature`, `func` requires %d arguments,'
+                ' but %s given' % (len(input_coredimss), len(args)))
+
+        args, dimsizess, loop_output_dims, outs, m_dims = self._get_args_transposed(  # NOQA
+            args, input_axes, input_coredimss,
+            outs, output_axes, output_coredimss)
+
+        # The output shape varies depending on optional dims or not
+        # TODO(ecastill) this only works for one out argument
+        out_shape = tuple([dimsizess[od][0] for od in loop_output_dims]
+                          + [dimsizess[od][0] for od in output_coredimss])
+        if outs is None:
+            outs = (cupy.zeros(out_shape, dtype=ret_dtype),)
+        elif outs[0].shape != out_shape:
+            raise ValueError(f'Invalid shape for out {outs[0].shape}'
+                             f' needs {out_shape}')
+        self._apply_func_to_inputs(0, dimsizess, loop_output_dims, args, outs)
+
+        tmp = outs
+        # This code credit goes to Dask
+        # https://github.com/dask/dask/blob/61b578f5a3ad88cbc6a8b9a73ce08c551bd969fa/dask/array/gufunc.py#L462-L503
+        # Treat direct output
+
+        if self._nout is None:
+            output_coredimss = [output_coredimss]
+
+        # Split output
+        # tmp might be a tuple of outs
+        # we changed the way we apply the function compared to dask
+        # we have added support for optional dims
+        leaf_arrs = []
+        for tmp in outs:
+            for i, (ocd, oax) in enumerate(zip(output_coredimss, output_axes)):
+                leaf_arr = tmp
+
+                # Axes:
+                if keepdims:
+                    slices = (len(leaf_arr.shape) * (slice(None),)
+                              + len(oax) * (numpy.newaxis,))
+                    leaf_arr = leaf_arr[slices]
+
+                tidcs = [None] * len(leaf_arr.shape)
+                for i, oa in zip(range(-len(oax), 0), oax):
+                    tidcs[oa] = i
+                j = 0
+                for i in range(len(tidcs)):
+                    if tidcs[i] is None:
+                        tidcs[i] = j
+                        j += 1
+                leaf_arr = leaf_arr.transpose(tidcs)
+                # Delete the dims that were optionals after the input expansion
+                if len(m_dims) > 0:
+                    shape = leaf_arr.shape
+                    # This line deletes the dimensions that were not present
+                    # in the input
+                    core_shape = shape[-len(ocd):]
+                    core_shape = tuple([
+                        d for d, n in zip(core_shape, ocd) if n not in m_dims])
+                    shape = shape[:-len(ocd)] + core_shape
+                    leaf_arr = leaf_arr.reshape(shape)
+                leaf_arrs.append(leaf_arr)
+        return (*leaf_arrs,) if self._nout else leaf_arrs[0]
```

Comment:
For the readabiltiy.
```suggestion
        return tuple(leaf_arrs) if self._nout else leaf_arrs[0]
```